### PR TITLE
Validity check v4 exomes

### PR DESCRIPTION
**Pull request by mike-w-wilson**
This unfurls and runs gnomad_methods validate_release_t on the v4 exomes release HT. It depends on https://github.com/broadinstitute/gnomad_methods/pull/625 and there is some room for improvements but I just wanted to get it in. It writes out the HT with unfurled annotations that will move to the next step.

### CODE REVIEW

**Comment by mike-w-wilson** [comment]
Loggers seem to not be working for me right now

**Comment by mike-w-wilson** [comment]
Loggers are a gnomad_methods issue: https://the-tgg.slack.com/archives/C6BD1M01E/p1697463923095569

File: gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+            #    logger.info("Filtering to test partitions...")
+            #    ht = filter_to_test(ht)
+
+            ht = prepare_ht_for_validation(ht)
+
+            validate_release_t(
+                ht,
+                subsets=SUBSETS,
+                pops=POPS,
+                monoallelic_expr=ht.info.monoallelic,
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py:439]: Also add ht.info.only_het?

File: gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+                verbose=args.verbose,
+                delimiter="_",
+                sample_sum_sets_and_pops={"non_ukb": POPS},
+                variant_filter_field="AS_VQSR",
+                problematic_regions=["lcr", "segdup", "non_par"],
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py:444]: Should probably also look at the other region flags 
```
        fail_interval_qc: bool, 
        outside_ukb_capture_region: bool, 
        outside_broad_capture_region: bool
```

File: gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+                single_filter_count=True,
+            )
+
+            # Note: Write saves time for the final export by not needing to run the
+            # VCF HT prep on each chromosome -- more needs to happen before ready for
+            # export, but this is an intermediate write.
+            logger.info("Writing prepared VCF HT for validity checks and export...")
+            ht.write(res.validated_ht.path, overwrite=True)
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py:452]: Checkpoint the prepared HT before running the validity checks?

File: gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+
+    for field in AS_VQSR_FIELDS:
+        vcf_info_dict[field] = t["vqsr_results"][f"{field}"]
+
+    # Add region_flag and allele_info fields to info dict
+    for field in ALLELE_TYPE_FIELDS:
+        vcf_info_dict[field] = t["allele_info"][f"{field}"]
+    for field in REGION_FLAG_FIELDS:
+        vcf_info_dict[field] = t["region_flag"][
+            f"{field}" if field != "nonpar" else "non_par"
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py:280]: Is this still needed, it looks like you changed it in `REGION_FLAG_FIELDS`

File: gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+# Drop decoy, still doesnt exist on 38
+REGION_FLAG_FIELDS = deepcopy(REGION_FLAG_FIELDS)
+REGION_FLAG_FIELDS = remove_fields_from_constant(
+    REGION_FLAG_FIELDS, ["decoy", "nonpar"]
+)
+REGION_FLAG_FIELDS.append("non_par")
+
+# Remove original alleles for containing non-releasable alleles
+ALLELE_TYPE_FIELDS = deepcopy(ALLELE_TYPE_FIELDS)
+MISSING_ALLELE_TYPE_FIELDS = ["original_alleles", "has_star"]
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py:59]: has_star is in the exomes

File: gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+    PipelineResourceCollection,
+    PipelineStepResourceCollection,
+)
+from gnomad_qc.v4.resources.basics import get_logging_path
+from gnomad_qc.v4.resources.release import release_sites, validated_release_ht
+
+# Add new site fields
+NEW_SITE_FIELDS = [
+    "monoallelic",
+    "singleton",  # NOTE: doesn't look like it was exported to VCF in V3.1 and is not part of gnomad_methods vcf module so maybe we drop it? Easy enough to look at AC?
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py:35]: Works for me

**Review by jkgoodrich — CHANGES_REQUESTED** [review]
I think this is looking good, I only looked at it for logic, which is totally fine at this point.


I think we might want to add in some other checks, like a check for length of globals being the expected length, and maybe counts being the expected counts. Maybe check for any uses of 'pop' or 'oth' in the globals?

Are there any other checks that we thing should be done on the globals? @ch-kr ?

Then only a few small suggestions

**Comment by mike-w-wilson** [comment]
I've updated this and it depends on https://github.com/broadinstitute/gnomad_methods/pull/630. In the `prepare_ht_for_validation` we select a subset of fields needed for VCF export and this gets checkpointed. I didnt want to add the freqs and fafs to this since they are so massive and are not needed downstream. For this reason, I put the global checks before that instead of calling it in gnomad methods validate_release_t.

**Comment by mike-w-wilson** [comment]
Back to you @jkgoodrich !

Test here: https://console.cloud.google.com/dataproc/jobs/4e803d67e7d04016a560a9c4bf069907/monitoring?region=us-central1&project=broad-mpg-gnomad

File: gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+from gnomad_qc.v4.resources.basics import get_logging_path
+from gnomad_qc.v4.resources.release import release_sites, validated_release_ht
+
+# TODO: Check global lengths match the fields they reference
+
+# Add new site fields
+NEW_SITE_FIELDS = [
+    "monoallelic",
+    "only_het",
+    "transmitted_singleton",
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py:43]: add sibling singletons

File: gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+from gnomad.utils.vep import VEP_CSQ_HEADER, vep_struct_to_csq
+
+from gnomad_qc.resource_utils import (
+    PipelineResourceCollection,
+    PipelineStepResourceCollection,
+)
+from gnomad_qc.v4.resources.basics import get_logging_path
+from gnomad_qc.v4.resources.release import release_sites, validated_release_ht
+
+# TODO: Check global lengths match the fields they reference
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py:37]:
```suggestion
```

File: gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+        pipeline_name="validate_and_export_vcf",
+        overwrite=overwrite,
+    )
+    validate_release_ht = PipelineStepResourceCollection(
+        "--validate-release-ht",
+        input_resources={
+            "release_ht": [
+                "gs://gnomad-tmp/gnomad.exomes.v4.0.qc_data/release/gnomad.exomes.sites.test.updated_101723.ht"
+            ]
+        },  # release_sites().ht()},
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py:141]: I know this is just for testing, but don't forget to change it back

File: gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py [hunk 10]
```diff
@@ -0,0 +1,550 @@
...
+    :param vep_csq_header: VEP CSQ header.
+    :return: Processed VEP CSQ header.
+    """
+    logger.info("Processing VEP CSQ header...")
+    vep_csq_header = vep_csq_header.split("|")
+    vep_csq_header = [f for f in vep_csq_header if f not in ["PolyPhen", "SIFT"]]
+    return vep_csq_header
+
+
+def check_globals_for_retired_terms(ht: hl.Table) -> None:
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/validate_and_export_exome_vcf.py:427]: This is great, thank you!!

**Review by jkgoodrich — CHANGES_REQUESTED** [review]
Only very small changes. I love seeing the output from the new checks, thank you so much for adding!!

File: gnomad_qc/v4/create_release/validate_and_export_vcf.py [hunk 11]
```diff
@@ -0,0 +1,543 @@
...
+    :param test: Whether to use test resources.
+    :return: Export resources.
+    """
+    export_pipeline = PipelineResourceCollection(
+        pipeline_name="validate_and_export_vcf",
+        overwrite=overwrite,
+    )
+    validate_release_ht = PipelineStepResourceCollection(
+        "--validate-release-ht",
+        input_resources={"release_ht": release_sites().ht()},
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/validate_and_export_vcf.py:136]:
```suggestion
        input_resources={"create_release_sites_ht.py": {"release_ht": release_sites()}},
```

File: gnomad_qc/v4/create_release/validate_and_export_vcf.py [hunk 12]
```diff
@@ -0,0 +1,543 @@
...
+    overwrite = args.overwrite
+    test = args.test
+    resources = get_export_resources(overwrite, test)
+
+    try:
+        if args.validate_release_ht:
+            logger.info("Running release HT validation...")
+            res = resources.validate_release_ht
+            res.check_resource_existence()
+            ht = res.release_ht
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/validate_and_export_vcf.py:476]:
```suggestion
            ht = res.release_ht.ht()
```

**Review by jkgoodrich — CHANGES_REQUESTED** [review]
two small things then lgtm

**Review by jkgoodrich — APPROVED** [review]
LGTM

**Pull request merged**